File: 99.recover-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def traversal(self, root):
        if not root:
            return
        self.traversal(root.left)
        if self.prev and root.val <= self.prev.val:
            if not self.first:
                self.first = self.prev
            self.second = root
        self.prev = root
        self.traversal(root.right)

    def recoverTree(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        self.first = None
        self.second = None
        self.prev = None
        self.traversal(root)
        tmp = self.first.val
        self.first.val = self.second.val
        self.second.val = tmp

File: 99.recover-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var recoverTree = function(root) {
    let first = null;
    let second = null;
    let prev = null;
    let traversal = function(root) {
        if (root == null) {
            return;
        }
        traversal(root.left);
        if (prev && root.val <= prev.val) {
            if (first == null) {
                first = prev;
            }
            second = root;
        }
        prev = root;
        traversal(root.right);
    }
    traversal(root);
    let tmp = first.val;
    first.val = second.val;
    second.val = tmp;
};